### PR TITLE
Use localhost address in conduit proxy within the controller pod

### PR DESCRIPTION
**Pull request by deebo91**
In order to reduce complexity with the conduit proxy in the controller pod. We want to use `127.0.0.1` in order to prevent undesired feedback loop effects when the controller pod proxy is routing traffic.

### CODE REVIEW

**Comment by briansmith** [comment]
I changed the base branch to PR #347 so that we can see the actual difference between the two. We can rebase the change on master afterward.

File: cli/cmd/inject.go [hunk 1]
```diff
@@ -201,6 +212,13 @@ func injectPodTemplateSpec(t *v1.PodTemplateSpec) enhancedPodTemplateSpec {
 		},
 	}
 
+	controlPlaneDNSName := fmt.Sprintf("tcp://proxy-api.%s.svc.cluster.local:%d", controlPlaneNamespace, proxyAPIPort)
+	if t.Namespace == controlPlaneNamespace && containsProxyAPIContainer(t.Spec, "proxy-api") {
```
Comment (briansmith) [inline cli/cmd/inject.go]: I don't think this is correct. We don't reserve the container name "proxy-api". In the future (if it doesn't already work now) we want to allow people to install the conduit control plane into the same namespace as their app, so we have to assume that there are other objects in the control plane namespace that we don't control. Some of them might have a container named `proxy-api` that is unrelated to our `proxy-api`.

Conversely, we already reserve the name `controller` in the conduit control plane namespace, so I personally think it's sufficient to just check that `t.Name == "controller"`. Regardless, checking `t.Name == "controller"` is at least necessary.

File: cli/cmd/inject.go [hunk 2]
```diff
@@ -201,6 +212,13 @@ func injectPodTemplateSpec(t *v1.PodTemplateSpec) enhancedPodTemplateSpec {
...
+		controlPlaneDNSName = fmt.Sprintf("tcp://%s:%d", "127.0.0.1", proxyAPIPort)
+	}
+	//remove namespace from container spec as its no longer needed to detect if we are processing a controller plane config
+	t.Namespace = ""
```
Comment (briansmith) [inline cli/cmd/inject.go]: I don't understand why adding and then removing a namespace is necessary.

File: cli/cmd/inject.go [hunk 3]
```diff
@@ -201,6 +212,13 @@ func injectPodTemplateSpec(t *v1.PodTemplateSpec) enhancedPodTemplateSpec {
 		},
 	}
 
+	controlPlaneDNSName := fmt.Sprintf("tcp://proxy-api.%s.svc.cluster.local:%d", controlPlaneNamespace, proxyAPIPort)
```
Comment (briansmith) [inline cli/cmd/inject.go]: 1. This is not a DNS name, it's a URL, so `controlPlaneDNSName` isn't the right name to use as the code is now.

2. We shouldn't duplicate the logic for formatting the "tcp://" prefix and and port number here and below. We should format the port number and the URL in one place.

File: cli/cmd/inject.go [hunk 4]
```diff
@@ -156,6 +158,15 @@ func injectDaemonSet(bytes []byte) (interface{}, error) {
 	}, nil
 }
 
+func containsProxyAPIContainer(spec v1.PodSpec, containerName string) bool {
```
Comment (adleong) [inline cli/cmd/inject.go]: I think this name isn't quite right.  This method seems to be a general way to test if a pod contains a named container, not just the Proxy API container.

Comment (deebo91) [reply cli/cmd/inject.go]: Thanks Alex, good suggestion. However, we no longer need this function with the new refactor in an upcoming commit. In this commit, we no longer need to search for the proxy container within a deployment spec. We can safely assume that if we read in the YAML object and the object is a deployment, we check for the deployment name to detect if it is the control plane. If it is, we can then pass in the DNS name override.

**Comment by briansmith** [comment]
>  Hairpin mode set to "promiscuous-bridge" but kubenet is not enabled, falling back to "hairpin-veth"
> So I enabled the kubenet plugin for minikube and that seemed to do the trick.

Please include an explanation of this issue (from discussion in #365) in a comment in the source code (not just the commit message) that explains why we are using "localhost" here.

Also in that comment, mention the other reasons we are doing this:
* When TLS is implemented, we don't want the internal traffic in the control plane to have the overhead of being to be encrypted and decrypted.
* We don't want one instance of a control plane pod to interact with other control plane pod instances. That is, we prefer to have each pod instance self-contained instance of trying to get them to load-balance each other.
* We don't want to collect metrics for internal traffic within the control plane.

**Comment by briansmith** [comment]
> Please include an explanation of this issue (from discussion in #365) in a comment in the source code (not just the commit message) that explains why we are using "localhost" here.

It would be great to reference kubernetes/minikube#1568 explicitly.

**Comment by deebo91** [comment]
Yep, I can definitely do that!

**Pull request merged**